### ISSUE TITLE
Tax levels depend on starting gold advantage

### ISSUE THREAD

**Issue by Heroesflorian** [Sh0wdown/UnofficialCrusaderPatch#382]
The ingame tax height depends on the AI/human advantage set in the start gold settings. 

Compare this (max. human advantage)...
![grafik](https://user-images.githubusercontent.com/12549394/57989753-e1e85800-7a9f-11e9-95aa-e663d2324b9f.png)
![grafik](https://user-images.githubusercontent.com/12549394/57989758-ef9ddd80-7a9f-11e9-8284-1bbf76c9b6fc.png)

...to this (max. AI advantage)...
![grafik](https://user-images.githubusercontent.com/12549394/57989777-325fb580-7aa0-11e9-8c93-c806b872fa11.png)
![grafik](https://user-images.githubusercontent.com/12549394/57989781-3be91d80-7aa0-11e9-9d24-523419d45e89.png)

...and this (no advantage for anyone):
![grafik](https://user-images.githubusercontent.com/12549394/57989791-60dd9080-7aa0-11e9-915d-2f652b3da6d3.png)
![grafik](https://user-images.githubusercontent.com/12549394/57989795-6b982580-7aa0-11e9-9d8e-79a131661049.png)

To save you further screenshots, here a small table summarizing things: 

Tax Level | No advantage | Slight advantage | Large advantage
------------ | ------------- | ------------- | -------------
Low Taxes (-2 popularity) | 0.6 GPM | 0.7 GPM | 0.8 GPM
...
High Taxes (-8 popularity) | 1.0 GPM | 1.2 GPM | 1.4 GPM
...
Cruelest Taxes (-24 popularity) | 2.0 GPM | 2.4 GPM | 3.0 GPM

(GPM = "Gold per Peasant per Month")

In other words, starting gold advantage can make your taxes 50% more effective. Starting gold amount is irrelevant, and starting gold disadvantage is treated as "no advantage". 
However, disadvantage for human players means advantage for the AI, and vice versa. 

This difference can greatly affect a match's outcome, especially considering AI behaviour - they will generally aim at higher taxes, preferring "-24 taxes" and extra rations over, for example, "-16 taxes", half rations and selling food. That is inefficient at best, but without tax rate advantage it is just completely ruinous... 

So why do I make this post? For two reasons: 
1. To spread the knowledge about this barely known game mechanic. 
2. To suggest some way of modifying that behaviour, possibly also in relation to what was mentioned in #327, to make different starting conditions more comparable and adjust tax balancing to personal taste.

**Comment by Krarilotus** [comment]
@Heroesflorian : Wow eye opening!
Tho my own testing for large advantage (in deathmatch 40k vs 3k) revealed, that in fact, low taxes yield 0.9 gpm in that mode and each tax tier above that another 0.3 gpm
Tested with 250 pop

Maybe there are even more differences. Maybe it is worth to investigate the full table! And also, are you sure about, that it applies for the ai the same way?

**Comment by ByBurton** [comment]
@Krarilotus Teste mal mit der Standardbevölkerung, wie @Heroesflorian

**Comment by Krarilotus** [comment]
> @Krarilotus Teste mal mit der Standardbevölkerung, wie @Heroesflorian

Yeah, das ist nen rundungsfehler, weil er die 0.3 numerisch als 0.299999999 darstellt, da 0.3 nicht im binärsystem genau dargestellt werden kann. Also ist mir klar wie er auf 0.8 gpm und 1.4 gpm kam. Der faktor zur normalen gpm ist ziemlich sicher 3/2. Und beim slight advantage entsprechend 5/4, das würde Sinn ergeben mit seinen Werten.

**Comment by GRhin** [comment]
by the way, Just went into the code with walrus grease, the equation for taxing is 
(tax_level - 1) / 5
this only occurs for actual taxing (tax_level 4+), low taxes is tax_level 4, and add 1 to the level per notch you increase the tax by. We didnt go into the bribe equation.
For gold advantage it works like this
Large human advantage = human gets tax amount * 1.5
Small human advantage = human gets tax amount * 1.2
Small AI advantage = AI get tax amount * 1.5
Large AI advantage = AI get tax amount * 2.5

**Comment by LordHansCapon** [comment]
Some decompiled code:

```
void __fastcall UpdateTaxing(void *param_1)

{
  int *playerGold;
  int currentTaxStep;
  int *playerHandle;
  int playerIndex;
  
  playerIndex = 1;
  playerHandle = (int *)((int)param_1 + 0x368bc);
  do {
    if ((0 < playerHandle[-0x83d]) && (0 < playerHandle[-0x7ed])) {
      currentTaxStep = *playerHandle;
      if (currentTaxStep < 3) {
        currentTaxStep =
             calculateBribeTaxIncomeForPlayer(param_1,playerIndex,currentTaxStep,playerHandle[-2]);
        playerHandle[-0x11] = playerHandle[-0x11] + currentTaxStep;
      }
      else {
        if (3 < currentTaxStep) {
          currentTaxStep = calculateTaxIncomeForPlayer(playerIndex,currentTaxStep,playerHandle[-2]);
          playerHandle[-0xc] = playerHandle[-0xc] + currentTaxStep;
        }
      }
      if (*(int *)((int)param_1 + 0x51a0c) != 0) {
        playerHandle[-0x11] = playerHandle[-0x11] / 10;
        currentTaxStep = playerHandle[-0xc] / 10;
                    /* Set player gold. */
        playerHandle[-0x71f] = playerHandle[-0x71f] + currentTaxStep;
        playerHandle[-0xc] = currentTaxStep;
        *(int *)(&playerGoldArray + playerIndex * 4) =
             *(int *)(&playerGoldArray + playerIndex * 4) + currentTaxStep;
        playerGold = playerHandle + -0x71f;
        *playerGold = *playerGold - playerHandle[-0x11];
        if (*playerGold < 0) {
          playerHandle[-0x71f] = 0;
        }
        playerHandle[-0x12] = playerHandle[-0x11];
        playerHandle[-0xd] = playerHandle[-0xc];
        playerHandle[-0xc] = 0;
        playerHandle[-0x11] = 0;
        *(undefined2 *)((int)playerHandle + 0xc2) = *(undefined2 *)(playerHandle + 0x30);
        *(undefined2 *)(playerHandle + 0x30) = *(undefined2 *)(playerHandle + -0x71f);
        playerHandle[-0x84d] = 0;
      }
    }
    playerIndex = playerIndex + 1;
    playerHandle = playerHandle + 0xe7d;
  } while (playerIndex < 9);
  return;
}
```

```
int calculateTaxIncomeForPlayer(int playerIndex,int currentTaxStep,int currentPeasants)

{
  int taxIncome;
  
  taxIncome = ((currentTaxStep + -1) * currentPeasants) / 2;
                    /* Is Skirmish */
  if (CurrentGameType != 0) {
                    /* Is NOT Player */
    if ((&PlayerArray)[playerIndex] == -1) {
                    /* Is AI */
      if ((&AIArray)[playerIndex] != 0) {
        if (AdvantageBalance == 4) goto LAB_004590f4;
        if (AdvantageBalance == 5) {
          return (taxIncome * 0xfa) / 100;
        }
      }
    }
    else {
      if (AdvantageBalance == 1) {
LAB_004590f4:
        return (taxIncome * 0x96) / 100;
      }
      if (AdvantageBalance == 2) {
        taxIncome = (taxIncome * 0x78) / 100;
      }
    }
  }
  return taxIncome;
}
```
```
int __thiscall calculateBribeTaxIncomeForPlayer(void *this,int playerIndex,int currentTaxStep,int currentPeasants)

{
  int taxIncome;
  
  taxIncome = ((5 - currentTaxStep) * currentPeasants) / 2;
  if (*(int *)(playerIndex * 0x39f4 + 0x3124c + (int)this) < 1) {
    taxIncome = 0;
  }
  return taxIncome;
}
```

**Comment by GRhin** [comment]
So bribe isnt effected by balance. and costs 1.5-2.5 gold per worker, ouch.
